File: response_gui.R
library(shiny)
library(dplyr)
library(readr)

# Define UI
ui <- fluidPage(
  titlePanel("Data Response Classifier"),

  sidebarLayout(
    sidebarPanel(
      fileInput("csv_file", "Choose CSV File",
                accept = c("text/csv", "text/comma-separated-values,text/plain", ".csv")),

      hr(),

      h4("Instructions:"),
      p("1. Upload a CSV with columns: id, name, value"),
      p("2. Review the data for each ID"),
      p("3. Click Positive or Negative"),
      p("4. Responses auto-save to result.csv"),

      hr(),

      textOutput("progress_text")
    ),

    mainPanel(
      h3(textOutput("current_id_text")),

      hr(),

      tableOutput("current_data_table"),

      hr(),

      fluidRow(
        column(6,
               actionButton("btn_positive", "Positive",
                          class = "btn-primary btn-lg btn-block",
                          icon = icon("thumbs-up"))),
        column(6,
               actionButton("btn_negative", "Negative",
                          class = "btn-danger btn-lg btn-block",
                          icon = icon("thumbs-down")))
      ),

      hr(),

      verbatimTextOutput("status_message")
    )
  )
)

# Define server logic
server <- function(input, output, session) {

  # Reactive values to store state
  rv <- reactiveValues(
    main_df = NULL,
    responses_df = NULL,
    processed_ids = c(),
    current_id = NULL,
    data_folder = NULL,
    result_file = "result.csv"
  )

  # Load existing result.csv if it exists
  load_existing_results <- function(folder_path) {
    result_path <- file.path(folder_path, rv$result_file)

    if (file.exists(result_path)) {
      tryCatch({
        existing_results <- read_csv(result_path, show_col_types = FALSE)

        if (all(c("id", "response") %in% colnames(existing_results))) {
          rv$responses_df <- existing_results
          rv$processed_ids <- existing_results$id
          return(TRUE)
        }
      }, error = function(e) {
        message("Warning: Could not load existing results: ", e$message)
      })
    }

    return(FALSE)
  }

  # Save to result.csv
  save_to_result_file <- function() {
    if (is.null(rv$data_folder)) return()

    result_path <- file.path(rv$data_folder, rv$result_file)

    tryCatch({
      write_csv(rv$responses_df, result_path)
    }, error = function(e) {
      showNotification(paste("Error saving to result.csv:", e$message),
                      type = "error")
    })
  }

  # Load next unprocessed ID
  load_next_id <- function() {
    if (is.null(rv$main_df)) return(FALSE)

    unique_ids <- unique(rv$main_df$id)
    unprocessed_ids <- setdiff(unique_ids, rv$processed_ids)

    if (length(unprocessed_ids) == 0) {
      showNotification("All IDs have been processed!", type = "message")
      rv$current_id <- NULL
      return(FALSE)
    }

    rv$current_id <- unprocessed_ids[1]
    return(TRUE)
  }

  # Handle CSV file upload
  observeEvent(input$csv_file, {
    req(input$csv_file)

    tryCatch({
      # Read the CSV
      df <- read_csv(input$csv_file$datapath, show_col_types = FALSE)

      # Validate columns
      required_cols <- c("id", "name", "value")
      if (!all(required_cols %in% colnames(df))) {
        showNotification(
          paste("CSV must contain columns:", paste(required_cols, collapse = ", ")),
          type = "error"
        )
        return()
      }

      rv$main_df <- df

      # Store the folder path
      rv$data_folder <- dirname(input$csv_file$datapath)
      # For user-selected files, use the directory where the file is located
      rv$data_folder <- dirname(normalizePath(input$csv_file$datapath))
      # Actually, for Shiny file uploads, save in current working directory
      rv$data_folder <- getwd()

      # Load existing results
      existing_found <- load_existing_results(rv$data_folder)

      if (existing_found) {
        num_existing <- length(rv$processed_ids)
        showNotification(
          paste0("Loaded ", nrow(df), " records\n",
                "Found existing result.csv with ", num_existing, " responses\n",
                "Filtering out already processed IDs..."),
          duration = 5,
          type = "message"
        )
      } else {
        # Initialize empty responses
        rv$responses_df <- tibble(id = numeric(), response = character())
        rv$processed_ids <- c()
        showNotification(paste("Loaded", nrow(df), "records"), type = "message")
      }

      # Load first ID
      load_next_id()

    }, error = function(e) {
      showNotification(paste("Error loading file:", e$message), type = "error")
    })
  })

  # Record positive response
  observeEvent(input$btn_positive, {
    req(rv$current_id)

    # Add response
    new_response <- tibble(id = rv$current_id, response = "positive")
    rv$responses_df <- bind_rows(rv$responses_df, new_response)
    rv$processed_ids <- c(rv$processed_ids, rv$current_id)

    # Save to file
    save_to_result_file()

    # Load next ID
    load_next_id()
  })

  # Record negative response
  observeEvent(input$btn_negative, {
    req(rv$current_id)

    # Add response
    new_response <- tibble(id = rv$current_id, response = "negative")
    rv$responses_df <- bind_rows(rv$responses_df, new_response)
    rv$processed_ids <- c(rv$processed_ids, rv$current_id)

    # Save to file
    save_to_result_file()

    # Load next ID
    load_next_id()
  })

  # Output: Current ID text
  output$current_id_text <- renderText({
    if (is.null(rv$current_id)) {
      "No ID loaded"
    } else {
      paste("ID:", rv$current_id)
    }
  })

  # Output: Current data table
  output$current_data_table <- renderTable({
    req(rv$main_df, rv$current_id)

    rv$main_df %>%
      filter(id == rv$current_id) %>%
      select(name, value)
  }, striped = TRUE, hover = TRUE, bordered = TRUE)

  # Output: Progress text
  output$progress_text <- renderText({
    if (is.null(rv$main_df)) {
      "No data loaded"
    } else {
      total_ids <- length(unique(rv$main_df$id))
      processed <- length(rv$processed_ids)
      paste0("Progress: ", processed, "/", total_ids, " IDs processed")
    }
  })

  # Output: Status message
  output$status_message <- renderText({
    if (is.null(rv$main_df)) {
      "Please upload a CSV file to begin."
    } else if (is.null(rv$current_id)) {
      paste0("All IDs processed!\n",
            "Responses saved to: ", file.path(rv$data_folder, rv$result_file))
    } else {
      paste0("Review the data above and click Positive or Negative.\n",
            "Auto-saving to: ", file.path(rv$data_folder, rv$result_file))
    }
  })
}

# Run the application
shinyApp(ui = ui, server = server)

File: response_gui.py
import pandas as pd
import tkinter as tk
from tkinter import ttk, messagebox, filedialog
import os


class ResponseGUI:
    def __init__(self, root):
        self.root = root
        self.root.title("Data Response Classifier")
        self.root.geometry("600x550")

        # Data storage
        self.df = None
        self.responses_df = pd.DataFrame(columns=['id', 'response'])
        self.current_id = None
        self.processed_ids = set()
        self.results_file = "result.csv"
        self.current_data_folder = None

        # Setup UI
        self.setup_ui()

    def setup_ui(self):
        # Title
        title_label = tk.Label(self.root, text="Data Response Classifier",
                               font=("Arial", 16, "bold"))
        title_label.pack(pady=10)

        # Load data button
        load_btn = tk.Button(self.root, text="Load CSV File",
                            command=self.load_data,
                            font=("Arial", 12),
                            bg="#4CAF50", fg="white", padx=20, pady=5)
        load_btn.pack(pady=10)

        # Display frame
        self.display_frame = tk.Frame(self.root, relief=tk.RIDGE, borderwidth=2)
        self.display_frame.pack(pady=20, padx=20, fill=tk.BOTH, expand=True)

        # ID label
        self.id_label = tk.Label(self.display_frame, text="ID: ",
                                 font=("Arial", 14, "bold"))
        self.id_label.pack(pady=10)

        # Data display (scrollable)
        self.data_text = tk.Text(self.display_frame, height=10, width=60,
                                font=("Arial", 11), state=tk.DISABLED)
        self.data_text.pack(pady=10, padx=10)

        # Buttons frame
        btn_frame = tk.Frame(self.root)
        btn_frame.pack(pady=10)

        # Positive button
        self.positive_btn = tk.Button(btn_frame, text="Positive",
                                     command=lambda: self.record_response("positive"),
                                     font=("Arial", 12), bg="#2196F3",
                                     fg="white", padx=30, pady=10,
                                     state=tk.DISABLED)
        self.positive_btn.pack(side=tk.LEFT, padx=10)

        # Negative button
        self.negative_btn = tk.Button(btn_frame, text="Negative",
                                     command=lambda: self.record_response("negative"),
                                     font=("Arial", 12), bg="#f44336",
                                     fg="white", padx=30, pady=10,
                                     state=tk.DISABLED)
        self.negative_btn.pack(side=tk.LEFT, padx=10)

        # Progress label
        self.progress_label = tk.Label(self.root, text="", font=("Arial", 10))
        self.progress_label.pack(pady=5)

        # Export button
        self.export_btn = tk.Button(self.root, text="Export Responses",
                                   command=self.export_responses,
                                   font=("Arial", 10), state=tk.DISABLED)
        self.export_btn.pack(pady=5)

    def load_existing_results(self, folder_path):
        """Load existing result.csv if it exists in the same folder"""
        result_path = os.path.join(folder_path, self.results_file)

        if os.path.exists(result_path):
            try:
                existing_results = pd.read_csv(result_path)
                if 'id' in existing_results.columns and 'response' in existing_results.columns:
                    self.responses_df = existing_results
                    self.processed_ids = set(existing_results['id'].tolist())
                    return True
            except Exception as e:
                print(f"Warning: Could not load existing results: {e}")

        return False

    def load_data(self):
        file_path = filedialog.askopenfilename(
            title="Select CSV File",
            filetypes=[("CSV Files", "*.csv"), ("All Files", "*.*")]
        )

        if not file_path:
            return

        try:
            self.df = pd.read_csv(file_path)

            # Validate columns
            required_cols = ['id', 'name', 'value']
            if not all(col in self.df.columns for col in required_cols):
                messagebox.showerror("Error",
                    f"CSV must contain columns: {', '.join(required_cols)}")
                return

            # Store the folder path
            self.current_data_folder = os.path.dirname(file_path)

            # Load existing results from result.csv if present
            existing_found = self.load_existing_results(self.current_data_folder)

            if existing_found:
                num_existing = len(self.processed_ids)
                messagebox.showinfo("Info",
                    f"Loaded {len(self.df)} records\n"
                    f"Found existing result.csv with {num_existing} responses\n"
                    f"Filtering out already processed IDs...")
            else:
                # Reset responses if no existing file
                self.responses_df = pd.DataFrame(columns=['id', 'response'])
                self.processed_ids = set()

            # Load first ID
            if not self.load_next_id():
                return

            # Enable buttons
            self.positive_btn.config(state=tk.NORMAL)
            self.negative_btn.config(state=tk.NORMAL)
            self.export_btn.config(state=tk.NORMAL)

            if not existing_found:
                messagebox.showinfo("Success", f"Loaded {len(self.df)} records")

        except Exception as e:
            messagebox.showerror("Error", f"Failed to load file: {str(e)}")

    def load_next_id(self):
        # Get unique IDs that haven't been processed
        unique_ids = self.df['id'].unique()
        unprocessed_ids = [uid for uid in unique_ids if uid not in self.processed_ids]

        if not unprocessed_ids:
            messagebox.showinfo("Complete",
                "All IDs have been processed!\n\n"
                "Don't forget to export your responses.")
            self.positive_btn.config(state=tk.DISABLED)
            self.negative_btn.config(state=tk.DISABLED)
            return False

        # Load next ID
        self.current_id = unprocessed_ids[0]
        self.display_current_id()
        return True

    def display_current_id(self):
        # Get data for current ID
        id_data = self.df[self.df['id'] == self.current_id]

        # Update ID label
        self.id_label.config(text=f"ID: {self.current_id}")

        # Update data display
        self.data_text.config(state=tk.NORMAL)
        self.data_text.delete(1.0, tk.END)

        # Display name and value for this ID
        for idx, row in id_data.iterrows():
            self.data_text.insert(tk.END, f"Name: {row['name']}\n")
            self.data_text.insert(tk.END, f"Value: {row['value']}\n")
            self.data_text.insert(tk.END, "-" * 50 + "\n")

        self.data_text.config(state=tk.DISABLED)

        # Update progress
        total_ids = self.df['id'].nunique()
        processed = len(self.processed_ids)
        self.progress_label.config(
            text=f"Progress: {processed}/{total_ids} IDs processed"
        )

    def save_to_result_file(self):
        """Automatically append to result.csv in the data folder"""
        if self.current_data_folder is None:
            return

        result_path = os.path.join(self.current_data_folder, self.results_file)

        try:
            self.responses_df.to_csv(result_path, index=False)
        except Exception as e:
            messagebox.showerror("Error", f"Failed to save to result.csv: {str(e)}")

    def record_response(self, response):
        if self.current_id is None:
            return

        # Add response to dataframe
        new_response = pd.DataFrame({
            'id': [self.current_id],
            'response': [response]
        })
        self.responses_df = pd.concat([self.responses_df, new_response],
                                     ignore_index=True)

        # Mark as processed
        self.processed_ids.add(self.current_id)

        # Automatically save to result.csv
        self.save_to_result_file()

        # Load next ID
        self.load_next_id()

    def export_responses(self):
        if self.responses_df.empty:
            messagebox.showwarning("Warning", "No responses to export!")
            return

        file_path = filedialog.asksaveasfilename(
            title="Save Responses",
            defaultextension=".csv",
            filetypes=[("CSV Files", "*.csv"), ("All Files", "*.*")]
        )

        if file_path:
            try:
                self.responses_df.to_csv(file_path, index=False)
                messagebox.showinfo("Success",
                    f"Responses exported to:\n{file_path}")
            except Exception as e:
                messagebox.showerror("Error",
                    f"Failed to export: {str(e)}")


def main():
    root = tk.Tk()
    app = ResponseGUI(root)
    root.mainloop()


if __name__ == "__main__":
    main()
